syncd/NotificationProcessor.cpp

```diff
@@ -397,7 +398,16 @@ void NotificationProcessor::process_on_port_state_change(
          * switch vid.
          */
 
-        oper_stat->port_id = m_translator->translateRidToVid(oper_stat->port_id, SAI_NULL_OBJECT_ID);
+        oper_stat->port_id = m_translator->translateRidToVid(rid, SAI_NULL_OBJECT_ID);
+
+        if(oper_stat->port_id == SAI_NULL_OBJECT_ID)
```

Comment:
if port_id is NULL then it will return NULL, if port_id is removed but it's not NULL, then translateRidToVid will  throw, since it will try to create new VID for null switch

what is actual scenario you are getting this error ? since if port was removed there should be no notification about this port any more

Reply:
port removal and port events (up/down) race condition (processed in separate thread). How on-fly (in transit) message being treated? SAI would have reported port event which got removed just after event has been raised. This is race condition. Need to drop this event, should not process.

Reply:
all notifications are synchronized with mutex that is uses to execute all api's, but yes notification arrives from vendor SAI internals in asynchronous way, so if  there is a waiting notification in the sai queue, to be posted, when another api is executed (like remove that port) then this notification will be processed under mutex, after api finished, so port will not be existing, and we will get error here.

so far i think we only hit this scenario once, in some wired case, but in general this is unavoidable, since notifications from SAI are designed to arrive in asynchronous way.

my proposal her would to to add method "tryTranslateRidToVid" which would return null when objects is not present, and log warning message with original RID value, and not send notification, or send notification with null port id just to OA awarness, which could be ignored then, since there is nothing that that could be done with that (better not send, and just log warn)

how about that ?

Reply:
tryTranslateRidToVid - would be good option from notification (or overload same function with try option) and send NULL object to OA. This gives an indication something being missed and can correlate (useful for debugging). Here is the trace which was seen. 

Nov 23 03:26:21.447281 sonic NOTICE swss#portsyncd: :- onMsg: Delete Ethernet24(ok) from state db
Nov 23 03:26:21.475279 sonic INFO syncd#syncd: [0] SAI_API_PORT:_brcm_sai_link_event_cb:432 Port 49 link down event cause: ADMIN_DOWN
Nov 23 03:26:21.475947 sonic NOTICE syncd#syncd: :- collectPortRelatedObjects: obtained 39 port oid:0x100000031 related RIDs
Nov 23 03:26:21.475947 sonic INFO syncd#syncd: [0] SAI_API_PORT:brcm_sai_remove_port:4531 REMOVE_PORT port_id:4294967345, bcm_port:49, phy:113
Nov 23 03:26:21.555213 sonic NOTICE swss#portsyncd: :- onMsg: nlmsg type:16 key:Ethernet16 admin:1 oper:0 addr:b8:6a:97:fd:d0:fa ifindex:148 master:0
Nov 23 03:26:21.555500 sonic NOTICE swss#portsyncd: :- onMsg: Publish Ethernet16(ok) to state db
Nov 23 03:26:21.555534 sonic NOTICE swss#portsyncd: :- onMsg: nlmsg type:16 key:Ethernet17 admin:1 oper:0 addr:b8:6a:97:fd:d0:fa ifindex:149 master:0
Nov 23 03:26:21.555843 sonic NOTICE swss#portsyncd: :- onMsg: Publish Ethernet17(ok) to state db
Nov 23 03:26:21.555843 sonic NOTICE swss#portsyncd: :- onMsg: nlmsg type:16 key:Ethernet18 admin:1 oper:0 addr:b8:6a:97:fd:d0:fa ifindex:150 master:0
Nov 23 03:26:21.556160 sonic NOTICE swss#portsyncd: :- onMsg: Publish Ethernet18(ok) to state db
Nov 23 03:26:21.606997 sonic NOTICE syncd#syncd: :- postPortRemove: removed 4 lanes from redis lane map for port RID oid:0x100000031
Nov 23 03:26:21.606997 sonic NOTICE syncd#syncd: :- postPortRemove: post port remove actions succeeded
Nov 23 03:26:21.607291 sonic ERR syncd#syncd: :- allocateNewObjectId: object type of switch oid:0x0 is SAI_OBJECT_TYPE_NULL, should be SWITCH

Reply:
will  you be willing to add tryTranslateRidToVid ?

Reply:
ok. let me add and push for review again. Thanks.